server/src/main/java/org/elasticsearch/action/SingleResultDeduplicator.java

```diff
@@ -45,50 +46,79 @@ public SingleResultDeduplicator(ThreadContext threadContext, Consumer<ActionList
 
     /**
      * Execute the action for the given {@code listener}.
-     * @param listener listener to resolve with execution result
+     * @param listener listener to resolve with execution result. The listener always has its threadContext preserved, i.e.
+     *                 when the listener is invoked, it will see its original threadContext plus any response headers generated
+     *                 by performing the {@code executeAction}.
      */
     public void execute(ActionListener<T> listener) {
         synchronized (this) {
-            if (waitingListeners == null) {
+            if (waitingListOfListenerWithContext == null) {
                 // no queued up listeners, just execute this one directly without deduplication and instantiate the list so that
                 // subsequent executions will wait
-                waitingListeners = new ArrayList<>();
+                waitingListOfListenerWithContext = new ArrayList<>();
             } else {
                 // already running an execution, queue this one up
-                waitingListeners.add(ContextPreservingActionListener.wrapPreservingContext(listener, threadContext));
+                waitingListOfListenerWithContext.add(
+                    new ListenerWithContext<>(
+                        ContextPreservingActionListener.wrapPreservingContext(listener, threadContext),
+                        // Only the first listener in queue needs the stored context which is used for running executeAction
```

Comment:
Likewise the first listener shouldn't need wrapping in a `ContextPreservingActionListener`?

But also if we're only capturing the thread context on the first listener, why not hold it in a separate field rather than holding a list of tuples? We're under a mutex here so can set both fields atomically.

Reply:
> Likewise the first listener shouldn't need wrapping in a ContextPreservingActionListener?

We still need wrap every listener, including the very first listener in `ContextPreservingActionListener`.

The trickly bit here is that we need the original context of a listener for two purposes:
1. We need to restore it when the listener is called with the computed result. This is done by wrapping every listener with `ContextPreservingActionListener`. Note that wrapping the very first listener is indeed a behaviour change of this PR. But I think it is a reasonable one. Previously we don't wrap the very first listener which means it could see mutated context due to computation while all other listeners will *not* see any mutated context even though computation is also done for some of them.
2. We need to restore the original context when we start the computation. This is necessary because the new batch of computation is started before the listener of last batch is called. Without restoring the original context, the new batch of computation will see the mutated context of the previous computation. This can cause multiple response headers to be accumulated which is something we want to fix with this PR. The comment says "Only the first listener in queue needs the stored context" because a single run of computation is done for a batch of listeners and we have to choose one of them for restore the original context. I believe this is [what you were suggesting](https://github.com/elastic/elasticsearch/pull/99728#discussion_r1332608467) as well. 

> if we're only capturing the thread context on the first listener, why not hold it in a separate field rather than holding a list of tuples? We're under a mutex here so can set both fields atomically.

You are right. The storedContext can be held in a separate field. I updated as suggested. Thanks!